### ISSUE TITLE
Don't mention constraints in errors when type parameters don't explicitly specify them

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#33436]
```
function f<T>(x: T) {
    x = {};
}
```

# Current

```
Type '{}' is not assignable to type 'T'.
  '{}' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.

Type '10' is not assignable to type 'T'.
  '10' is assignable to the constraint of type 'T', but 'T' could be instantiated with a different subtype of constraint '{}'.
```

Nobody ever mentioned a constraint or `{}` in this code, so why is it being displayed? As a user, this is telling me more than I needed or wanted to know.

# Proposed

```
Type '{}' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to '{}'.


Type '10' is not assignable to type 'T'.
  'T' could be instantiated with an arbitrary type which could be unrelated to '10'.
```

**Comment by dagda1** [comment]
I see the same error in this [playground](https://typescript-play.js.org/#code/PQKhCgAIUhlAXAhgOwCaIDYHtkFNK4AeiAtgA4b5YBmkAKgJ5m4AKATlmQM6TwAWieJADGKSACN8bAK7JIASzmJILDIgYBzDrNRQI0SAEFhAa2RYA7pVQbcJXMngAuPZAC0kAALoN6AIyQ1FhsvHz4XKQU+MJYqPj88jwWiDzB8hqKmBgMEim4qJA4kCgF-Pjycco0xYGywvDBhQBuuCEABnQA-G2uHp4kKYi4+dKBjWWQEeSUIrHxWLy4XEITmBpp-CRcvV7IiOj2o0EhYRhkkBby-KHRwfhxXOnIesDge-ZcZIjC+GRYwpAAN5QSCg4DASAAMTqDTYINBingrWo33wCEE8mE0OQ9WCAB4AOIAPiB8NB5IGZDxkIIhCRaB4BIcrUxhKJABpIABVIkACjJ5MFvDYKC4xxITkgvMQkpYiBF9iRbC41KJAG0AAwAXQAlJAALwkrnsgVC0GUr7iSiSyGm0E6yVMvBsVmQzlqrlaokAbjJAF9wGTwZAAKLEaZLMkxZDLYpsEUMbG4tiS9HwTFJ2F4lAMNVeg2ks3mxBkSV4wycgBCfOoyEl0slhj1hsglc5KMbeYdrbzBpJwKLQrYuHg0jYchRADpKbzazrfYOA4K-QvydHY1hxAArXD1TPBVNIdNYmH4zc7+ok-WFouUssV1s1utSmVGZsktuBV9NyWVvs3wdQWHUdx0CZBZ0Qec7UgJdyRXKMcFjZBpAwNQrVwfcUzgI8M1PNg8XPXchAAH0gZDUMgUidFwahFHyK8AKFO9IHLKs+WgwVa3rb931bE1APJDsjEosiUIwETqNovBdEHbs-1I8jxKotAaLogoWwHATIGAscJ2UABCa9JLUyBOjAiC9UlFFVzNWDQXghCYyEVAsGkdCQmvXlCElZCSEkNheMIAwACZfUc2NCB4a95QTTDpxLXkXLcyg2HdAAWTlgt1b1IGDSKfOkPzWjzcKhCC69CL3PD4rIRLXPczk0uC+dcohbzRKKuFBXXMqADkxILRTEHQuKZyShrRNQlq8v61CCtQ0rIEIWAsHsQaxOGyhRoS8aUsa4oeF8-yJJUqT8mmtqVvsArOpOuIzt0RbqDww8MRPHEs1kMxLDkUiczzBjNLNGda3bSDGMHHTQMMeN1EnRIYYTCyOKLMyYvUbbatBr8dRRs0rIMozTrUvHALMyr4Ex2dkDB3GtMA+a0K26qQZpnGbKFOyYLC7rEKEBgotqD7ghqurktadLMuy1rIAFm7-JK3mnNlgtnuFthRd2iXICai7Zfl1pFoYWbxOvNXk01+q9smjA9eNsTGaNq78DNlmdqt7WmoOjrjuU+61Lt52DZCP3VOkwNBWDOgliEVz4DIaR4AjtdEKwShJ2wDReQAIjoQwWEgFplXkIo-AAZmzqCla4NPcAzrAs+zgBiegmFYDhuBUDhqmqABhHAfjIeBK459da-rxu-EnScAA4R+T0Ex-TzP+UHSLNS1A19WvGfIAAMj3pauDVPxN+3680tMyBs6wExIACdGckpbPIElbPzFju+H9hp+S2zskq4pxjOPFe5Vt6QBnpfMyN877BUKNuIikBn6v2vh-QosD4EXiEM-QBi9U7Lwbl5E2W9ryKSvjAyAZcbZIL-ig9+WBP6UOoTg0e+C66gOdiQiB5Db460mKtfAyC35oN4ZfGua0WHhRAYQ6CAsN5cN3gfWWx9T5cLSqTaBvCACskBbDOkxHGBMNCyD-3psIhh6DIA6L0SyAEj9jGmPJLg2YwCCFZ1kQoqB19eEADZdHMhdACCmDi6EiLvn4mxgTMGIOfgA1hrj2EyMHPbCi59qGaLvgAdn8fogEZChGoIsbw7JkSDH5L-nEqRbjV5FgYJwtJkCNHeLvrvUpeSNroQcaTcxjDWkBLKR0mYsTBRQSDBCVQhVxCKA0GSREyJUSt2YOwTgKo6AFkBH6TkcoFQ8CIPSVAPBEbqAADLyBMLgbMyAGAMTwEXfs0FH6SiBoBRQ1BWiSjWXshwByjA-zxK81oRgSRmTVIYTePlcBFw5oOHqMh6iynlKQY+2o8zQrgmi0EiknmkwBVhMhXyGT0CvmqW5rRwVkUha0DFQpYXSHhTbal8FBzGWkti+muLJQsvyLSfZPA1kgtJWwclgrqXV3gHC5wkAuWoEZdS2QQQdBsq0hyyAapeo4BNptC5dAiRalFUA5YEqEU7I3rK-0C9eBt0gAAWUEMIPgeIdUFk+XSb5PBcXcmgmZXkGyeVus9fTUEaMrmk1BB8tuyzuB4h5GqMMwgMDSDiHiUNRZzkMGqIwJZHcuD8UDWabO8rXJoEcXm3Vaps64uzrqP1hLBVNILcgBVxbukQ0DWqAA0rgHIihQyEHjYmi5aaM0RuzZyBtTbUDZ11eGrNKzk15sHHQFNgp1SdoYFqctlbN4Ep+XWhdqMKVF2XUWSUa6U1+ljX2hNSah20Eze3FZY7C06CnVqaCEKoUWvgFa7ZpARytFWQxe9kbANqltfAe1jqy0Vsba0KtPNyTfuYJAJ0tjHVbMRVsGtPyjkMFOecy51yCy4fwxcnMRIgMjrndBOguazS-q2GSdU4HIM6o3dnWl9R4PgADEAA).  I am at a loss to what the error is and where `{}` is coming from

**Comment by jack-williams** [comment]
Looks like the default constraint is coming through as `{}`, but shouldn't it be `unknown` now?

**Comment by dagda1** [comment]
i’m seeing this in many places in many different projects

**Comment by dagda1** [comment]
i have no idea what the error message means

**Comment by fatcerberus** [comment]
> Looks like the default constraint is coming through as {}, but shouldn't it be unknown now?

I had the same question.  It was changed to `unknown` in TS 3.5 but still shown as `{}` in (some? all?) relevant error messages.

**Comment by jack-williams** [comment]
I debugged this and the summary is that `noConstraintType` is defined as an empty object type.

```ts
const noConstraintType = createAnonymousType(undefined, emptySymbols, emptyArray, emptyArray, undefined, undefined);
```

Changing this type to be defined as `unknown` changes the error message accordingly.

Linking to this issue for my own sake #32814 as I think they are both affected by the definition of `noConstraintType`.

**Comment by jack-williams** [comment]
On the topic of error messages: A proposal:
- `'T' could be instantiated with a type that is unrelated to '10'.`,
- or, just specify no additional elaboration.

**Comment by jack-williams** [comment]
@dagda1 Sorry, to answer you questions.

Assigning a concrete type to a generic type parameter is incorrect because the type parameter can always be instantiated to some arbitrary type:
```ts
function f<T>(x: T) {
    x = {};
}
f(3);
f(true);
f("something else");
```
In the body of the function you have no control over the instantiation by the calling context, so we have to treat things of type `T` as opaque boxes and say you can't assign to it.

A common mistake or misunderstanding was to constraint a type parameter and then assign to its constraint, for example:
```ts
function f<T extends boolean>(x: T) {
    x = true;
}
f<false>(false);
```
This is still incorrect because the constraint only restricts the upper bound, it does tell you how precise `T` may really be.

To try and address this misunderstanding the additional error elaboration was provided, but in the most degenerate case, an unconstrained parameter, the elaboration is really just noise.

**Comment by dagda1** [comment]
@jack-williams thank you for the explanation although I see this error a crazy amount of times now and the implicit `{}` constrain is what I always see.

The example in [this playground](https://typescript-play.js.org/#code/PQKhCgAIUhlAXAhgOwCaIDYHtkFNK4AeiAtgA4b5YBmkAKgJ5m4AKATlmQM6TwAWieJADGKSACN8bAK7JIASzmJILDIgYBzDrNRQI0SAEFhAa2RYA7pVQbcJXMngAuPZAC0kAALoN6AIyQ1FhsvHz4XKQU+MJYqPj88jwWiDzB8hqKmBgMEim4qJA4kCgF-Pjycco0xYGywvDBhQBuuCEABnQA-G2uHp4kKYi4+dKBjWWQEeSUIrHxWLy4XEITmBpp-CRcvV7IiOj2o0EhYRhkkBby-KHRwfhxXOnIesDge-ZcZIjC+GRYwpAAN5QSCg4DASAAMTqDTYINBingrWo33wCEE8mE0OQ9WCAB4AOIAPiB8NB5IGZDxkIIhCRaB4BIcrUxhKJABpIABVIkACjJ5MFvDYKC4xxITkgvMQkpYiBF9iRbC41KJAG0AAwAXQAlJAALwkrnsgVC0GUr7iSiSyGm0E6yVMvBsVmQzlqrlaokAbjJAF9wGTwZAAKLEaZLMkxZDLYpsEUMbG4tiS9HwTFJ2F4lAMNVeg2ks3mxBkSV4wycgBCfOoyEl0slhj1hsglc5KMbeYdrbzBpJwKLQrYuHg0jYchRADpKbzazrfYOA4K-QvydHY1hxAArXD1TPBVNIdNYmH4zc7+ok-WFouUssV1s1utSmVGZsktuBV9NyWVvs3wdQWHUdx0CZBZ0Qec7UgJdyRXKMcFjZBpAwNQrVwfcUzgI8M1PNg8XPXchAAH0gZDUMgUidFwahFHyK8AKFO9IHLKs+WgwVa3rb931bE1APJDsjEosiUIwETqNovBdEHbs-1I8jxKotAaLogoWwHATIGAscJ2UABCa9JLUyBOjAiC9UlFFVzNWDQXghCYyEVAsGkdCQmvXlCElZCSEkNheMIAwACZfUc2NCB4a95QTTDpxLXkXLcyg2HdAAWTlgt1b1IGDSKfOkPzWjzcKhCC69CL3PD4rIRLXPczk0uC+dcohbzRKKuFBXXMqADkxILRTEHQuKZyShrRNQlq8v61CCtQ0rIEIWAsHsQaxOGyhRoS8aUsa4oeF8-yJJUqT8mmtqVvsArOpOuIzt0RbqDww8MRPHEs1kMxLDkUiczzBjNLNGda3bSDGMHHTQMMeN1EnRIYYTCyOKLMyYvUbbatBr8dRRs0rIMozTrUvHALMyr4Ex2dkDB3GtMA+a0K26qQZpnGbKFOyYLC7rEKEBgotqD7ghqurktadLMuy1rIAFm7-JK3mnNlgtnuFthRd2iXICai7Zfl1pFoYWbxOvNXk01+q9smjA9eNsTGaNq78DNlmdqt7WmoOjrjuU+61Lt52DZCP3VOkwNBWDOgliEVz4DIaR4AjtdEKwShJ2wDReQAIjoQwWEgFplXkIo-AAZmzqCla4NPcAzrAs+zgBiegmFYDhuBUDhqmqABhHAfjIeBK459da-rxu-EnScAA4R+T0Ex-TzP+UHSLNS1A19WvGfIAAMj3pauDVPxN+3680tMyBs6wExIACdGckpbPIElbPzFju+H9hp+S2zskq4pxjOPFe5Vt6QBnpfMyN877BUKNuIikBn6v2vh-QosD4EXiEM-QBi9U7Lwbl5E2W9ryKSvjAyAZcbZIL-ig9+WBP6UOoTg0e+C66gOdiQiB5Db460mKtfAyC35oN4ZfGua0WHhRAYQ6CAsN5cN3gfWWx9T5cLSqTaBvCACskBbDOkxHGBMNCyD-3psIhh6DIA6L0SyAEj9jGmPJLg2YwCCFZ1kQoqB19eEADZdHMhdACCmDi6EiLvn4mxgTMGIOfgA1hrj2EyMHPbCi59qGaLvgAdn8fogEZChGoIsbw7JkSDH5L-nEqRbjV5FgYJwtJkCNHeLvrvUpeSNroQcaTcxjDWkBLKR0mYsTBRQSDBCVQhVxCKA0GSREyJUSt2YOwTgKo6AFkBH6TkcoFQ8CIPSVAPBEbqAADLyBMLgbMyAGAMTwEXfs0FH6SiBoBRQ1BWiSjWXshwByjA-zxK81oRgSRmTVIYTePlcBFw5oOHqMh6iynlKQY+2o8zQrgmi0EiknmkwBVhMhXyGT0CvmqW5rRwVkUha0DFQpYXSHhTbal8FBzGWkti+muLJQsvyLSfZPA1kgtJWwclgrqXV3gHC5wkAuWoEZdS2QQQdBsq0hyyAapeo4BNptC5dAiRalFUA5YEqEU7I3rK-0C9eBt0gAAWUEMIPgeIdUFk+XSb5PBcXcmgmZXkGyeVus9fTUEaMrmk1BB8tuyzuB4h5GqMMwgMDSDiHiUNRZzkMGqIwJZHcuD8UDWabO8rXJoEcXm3Vaps64uzrqP1hLBVNILcgBVxbukQ0DWqAA0rgHIihQyEHjYmi5aaM0RuzZyBtTbUDZ11eGrNKzk15sHHQFNgp1SdoYFqctlbN4Ep+XWhdqMKVF2XUWSUa6U1+ljX2hNSah20Eze3FZY7C06CnVqaCEKoUWvgFa7ZpARytFWQxe9kbANqltfAe1jqy0Vsba0KtPNyTfuYJAJ0tjHVbMRVsGtPyjkMFOecy51yCy4fwxcnMRIgMjrndBOguazS-q2GSdU4HIM6o3dnWl9R4PgADEAA) has nothing to do with assigning to a constraint and I would really appreciate an explanation of what the error message is trying to tell me here.

I have so many other example.

**Comment by jack-williams** [comment]
Line 34 is trying to assign `A | B | null | undefined` to `B | null | undefined` which fails because `A` is not related to any union constituent in `B | null | undefined`. Type parameter `A` does not have an explicit constraint which is why you see the additional spurious message; the error message is using the implicit constraint when it should not be and that should be getting fixed.

The error message is correct though, and symptomatic of the fact that `nullableFunctor` does not satisfy the functor laws. For the correct typing you need negation types and to constraint `A` to `not (null | undefined)`.

**Comment by theseyi** [comment]
@jack-williams I semi understand why this change was made, but it's not clear how to resolve it. Most of the documentation in this repo seem to say the same thing, but the error does not give a suggestion on how to resolve this.
I have a simple playground with comments on the issues I'm facing with this, appreciate your help [here](https://www.typescriptlang.org/play/#code/MYGwhgzhAECC0G8BQ1XQLYFMAuALA9gCawA8AKtJgB7aYB2hMdArugEaYBOAfABR0AuVGQCUQisjRToAehnQAEvgDu0QvmgBJVACtmEbNE45mnOtDDQW7LgBojJs9AAsAbhTTUx7KfPOLMGTuUgC+SGFIoJAwAEKUNPSMcJJoctAA6rgAntDAYObadGCcnCrQeJjlWQAOlfgAZlbQbJCYhND45hW5nQacYACWdIZgbPgAbpgeqFh4RLD8QtYcnGJWrCuI09LevtAArMFoYSFAA)

**Comment by jack-williams** [comment]
@theseyi It's not clear to me why `A.methodA` needs to be a generic method. The error is correct though:

Every `B` should be a subtype of `A`, so I should always be able to assign a `B` to an `A`, but the type of `B.method` is _weaker_. For example:
```ts
const a: A = new B();
const three: 3 = a.methodA(3);
```
Because `A.methodA` is generic I think I'm getting `3`, but I'm really calling `B`'s method, which returns 5.

It's hard for me to help you fix this because I'm not sure what you're trying to achieve, but having a parent method return a generic type, and a child method return a non-generic type, is usually wrong.

**Comment by theseyi** [comment]
Thanks for responding, I agree in attempting to get to the salient  parts of the issue, I may have fuzzied the example and concrete values make it unclear.
Hopefully, this is a more vivid example, TS does not complain, but is this the correct approach [here](https://www.typescriptlang.org/play/#code/JYOwLgpgTgZghgYwgAgEJwM4QKIDcLjIDeAUMuchPuAHJwC2EAXMhmFKAOYDcZFwAExYgArvQBG0XgF8SJUJFiIUAYRFsA9vTwEwAFQCeABwgAePQD5KAD0ggBGNJhzUwxPuQTqwWgCJwwOAB+Fj0SWTkEABtMRx1wAAk4eyjocxs7BycseLArUgpKVxDkPV5CmGAoF10ACipdOkYWNg4QTgAaZEFhMUkoAEp3QsKEDRA2It1kAF5kMAALYAwAOnERYCiBXPrXJoguwQHykc9xjA1UlaiNTl3dY49kWUL1ze3XUxULe9oGZlY7C4hyEyFEEmgXQA+hojGBgOcSioBix0DlXMNTsgxhM3OJnLlUQSMXMCljCg0-owOk9yYJaRRpLwGeRqmARFAQMh8ejdCdyLIItFYsg1Jp6DQ+tBckkUtAMgQsjLkgJUlAvt4tLlDCZTOD+hZ8k83ltcuYfpSwPsWkD2iDehCoF0vOL-IFQijRZrtK4dWZLJjTjjJjyaoQ5hgRCYoGsNqbXL8rf8QY9ycG3C6fD7dCwxVntcZ-VZScgVmXQ7lnd63XBnszyWyOVzM1rXPznuEgA)?

In other words, I have a class that is generic over a base type (because as you mention inherited and base methods have to be generic), and a subclass that is generic over a more specialized extension of that base type.

**Comment by jack-williams** [comment]
@DanielRosenwasser I have a fix (#33445) that removes the useless error message. Is the proposed message final and OK to go ahead and add?
